aws/rust-runtime/aws-sigv4/src/http_request/uri_path_normalization.rs

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::borrow::Cow;
+
+// Normalize `uri_path` according to
+// https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
+pub(super) fn normalize_uri_path(uri_path: &str) -> Cow<'_, str> {
+    // If the absolute path is empty, use a forward slash (/).
+    if uri_path.is_empty() {
+        return Cow::Borrowed("/");
+    }
+
+    // The canonical URI is the URI-encoded version of the _absolute_ path component of the URI.
+    let result = if uri_path.starts_with('/') {
+        Cow::Borrowed(uri_path)
+    } else {
+        Cow::Owned(format!("/{uri_path}"))
+    };
+
+    if !result.contains('.') {
+        return result;
+    }
+
+    Cow::Owned(normalize_path_segment(&result))
+}
+
+// Implement 5.2.4. Remove Dot Segments in https://www.rfc-editor.org/rfc/rfc3986
+//
+// The function assumes that `uri_path` is an absolute path,
+// starting with a forward slash.
+fn normalize_path_segment(uri_path: &str) -> String {
+    let number_of_slashes = uri_path.matches('/').count();
+    let mut normalized: Vec<&str> = Vec::with_capacity(number_of_slashes + 1);
+
+    for segment in uri_path.split('/') {
+        match segment {
+            "." => {}
+            ".." => {
+                normalized.pop();
+            }
+            otherwise => normalized.push(otherwise),
+        }
+    }
+
+    let mut result = normalized.join("/");
+
+    // Even though `uri_path` starts with a `/`, that may not be the case for `result`.
+    // An example of this is `uri_path` being "/../foo" where the corresponding `result`
+    // will be "foo".
+    if !result.starts_with('/') {
+        result.insert(0, '/');
+    }
```

Comment:
I think `result` will never start with `/` in this code, so the `result.insert` will always run.

Reply:
Interestingly, `result` starts with `/` most of the time. For instance, if the input to the function is `/a/b/c`, then the vector `normalized` will contain `["", "a", "b", "c"]`. After `normalized.join`, `result` will then be `/a/b/c`.

Reply:
Ah, yeah, you're right. I forgot about the leading `/` causing a `""` in the split.